src/Aframe/volume.js

```
@@ -267,68 +269,99 @@ AFRAME.registerComponent("volume", {
+
+    // Resize viewport to scene
     uniforms.viewPort.value = new Vector2(
-      this.canvas.width,
-      this.canvas.height
+      this.scene.canvas.width,
+      this.scene.canvas.height
     );
 
-    uniforms.dim.value = dim;
-    uniforms.zScale.value = zScale;
-
-    uniforms.blending.value = blending.blending;
+    // Apply model properties
     uniforms.intensity.value = intensity;
-    uniforms.slice.value = slices;
-    uniforms.useLut.value = useTransferFunction;
+    uniforms.use_lut.value = useTransferFunction;
+    uniforms.u_data.value = texture; // Model data
+    uniforms.u_lut.value = transferTexture; // Color Map + transfer function
+
+    // blending, sliders, slices, and spacing are updated separately
+    const otherUniforms = this.getMesh().material.uniforms;
+    uniforms.blending.value = otherUniforms.blending.value;
+    uniforms.box_min.value = otherUniforms.box_min.value;
+    uniforms.box_max.value = otherUniforms.box_max.value;
+    uniforms.slices.value = otherUniforms.slices.value;
+    uniforms.dim.value = otherUniforms.dim.value;
+    uniforms.zScale.value = otherUniforms.zScale.value;
+    return uniforms;
+  },
+
+  updateBlending: function () {
+    const { blending } = this.data;
+    const uniforms = this.getMesh().material.uniforms;
+    uniforms.blending.value = blending;
+  },
+
+  updateSlices: function () {
+    const { slices } = this.data;
+    const uniforms = this.getMesh().material.uniforms;
 
-    uniforms.u_data.value = texture;
-    uniforms.u_lut.value = transferTexture;
+    uniforms.slices.value = slices;
+    uniforms.dim.value = Math.ceil(Math.sqrt(slices));
+  },
+
+  updateSpacing: function () {
+    const { spacing } = this.data;
+    const uniforms = this.getMesh().material.uniforms;
+    const texture = uniforms.u_data.value; // Image
+    const dim = uniforms.dim.value;
+    const slices = uniforms.slices.value;
+
+    const volumeScale = texture
```

this is kind of big for a ternary - maybe make an if/else?

---

or get the width/height based on texture and then make the `new Vector3` once outside of control flow with those computed values

---

`texture` is undefined on the initial render. Looks like it's safe to just encapsulate everything in `if(texture) {}` because `this.updateSpacing` is called whenever the models update

---

This does get me thinking though. Should we enforce that `texture.image.width` and `texture.image.height` are the same for every `model` in `models`? I can only image this will become an issue when we blend the textures together. 

I imagine tiff to png outputs the same sized images but just in case... or is it safe to just let it error out if that becomes a problem?